internal/pipeline/fn.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pipeline
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"go.uber.org/zap"
+	"sigs.k8s.io/kustomize/kyaml/comments"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	// annotation used to track resources for preserving comments during the function execution
+	idAnnotation = "config.k8s.io/id"
+)
+
+// KRMFn defines capabilities of a KRM function. It primarily knows
+// how to run itself. It accepts input from an io.Reader that is
+// a stream of KRM objects in the form of KRM FunctionSpec. It also accepts
+// a io.Rriter where it writes the result of running a KRM function.
+type KRMFn interface {
+	Run(r io.Reader, w io.Writer) error
+}
+
+// So we will have different types of KRMFn implementation:
+// ExecFn: that can execute a binary on local machine
+// NetworkFn: that can invoke a endpoint remotely
+// ContainerFn: that can execute a container locally
+// Built-in functions that are compiled into kpt
+
+// annotator is built kpt function for annotating KRM resources.
+type annotator struct {
+	key    string
+	value  string
+	logger *zap.SugaredLogger
+}
+
+func (a *annotator) Run(r io.Reader, w io.Writer) error {
+	rw := &kio.ByteReadWriter{
```

Comment:
Based on just this example, why isn't `rw' passed as an argument to this method?

Reply:
If the function implementation always creates a rw from r and w and operates on that, that's a signal that we should just be passing rw as an argument?